### ISSUE TITLE
Support decoding WhatsApp Status changes

### ISSUE THREAD

**Issue by gisms** [sepinf-inc/IPED#616]
I'm indexing some .ufdr files, so far only iphones, and I've noticed less conversations coming up on iped. I didn't get to analyze it in depth, but in general it seems to me that the missing conversations are those that only have media attached, or a link sent, without any text input, without even that standard system text input. Did you experience something similar?

**Comment by lfcnassif** [comment]
Be aware the internal parser needs sha-256 to be enabled and it does not recover deleted messages for now, please check this https://github.com/sepinf-inc/IPED/issues/309#issuecomment-721463990

and please report back if that's not the case.

**Comment by gis-msouza** [comment]
Wow, I looked it up on the forum before, but I think the name of the comment you linked wasn't as directed to my case, although it helped me.

I generated the ufdr report including 256 and md5 and also used ipedconfig with both.

- External parser: actually brought all chats.

- internal parser still missing some chats as explained

**Comment by lfcnassif** [comment]
@fmpfeifer any idea?
@gis-msouza could you share privately the WA database to reproduce this?

**Comment by fmpfeifer** [comment]
> @fmpfeifer any idea?
> @gis-msouza could you share privately the WA database to reproduce this?

I never came across something like that. A Chatstorage.sqlite that triggers the bug would be really useful to undestand what is going on.

**Comment by lfcnassif** [comment]
> * Internal parser: still missing some chats as explained

@gis-msouza just to confirm, are those missing messages or chats flagged as deleted in external parser output?

**Comment by gis-msouza** [comment]
apologies for the delay.

cellebrite did not bring deletion information. The chat is not marked as deleted. But, I managed to confront in reality and is without the messages.

For greater visualization of the scenario, just out of curiosity, I took a look at ChatStorage, there is an entry in ZWACHATSESSION with reference to the contact who sent the conversations.
There are the messages entries in the ZWAMESSAGE, but no reference to the contact who sent them, ZFROMJID status@broadcast.

In ZWAMEDIAITEM, the fields ZMEDIAURL and ZXMPPTHMBPATH and ZMETADATA are filled in and in fact only the thumbs are are available in the ufdr.

**Comment by lfcnassif** [comment]
> There are the messages entries in the ZWAMESSAGE, but no reference to the contact who sent them, ZFROMJID status@broadcast

This is a good tip. Please keep this Chatstorage.sqlite so you could help us to test a possible future solution to this if you cannot share the database privately.

@fmpfeifer if you have time (or not) to take a look at this in the near future, please let me know.

**Comment by gis-msouza** [comment]
> > There are the messages entries in the ZWAMESSAGE, but no reference to the contact who sent them, ZFROMJID status@broadcast
> 
> This is a good tip. Please keep this Chatstorage.sqlite so you could help us to test a possible future solution to this if you cannot share the database privately.
> 
> @fmpfeifer if you have time (or not) to take a look at this in the near future, please let me know.

**Comment by gis-msouza** [comment]
Yes it is possible for me to test again any time. Let me know.

Ps: sorry i closed by mistake.

**Comment by fmpfeifer** [comment]
I think I can take a look at it in a few days.
I think that this status@broadcast is used for the whatsapp status messages, and as far as I can remember, we are not decoding those messages yet.

**Comment by lfcnassif** [comment]
@gis-msouza could you confirm the missing messages refer just to status changes? What thumbs are available in the UFDR? Contact thumbs or some thumb referenced by the missing messages?

**Comment by gisms** [comment]
yes, they are chats with id filled in with phone@status. so that's the reason. the images are not contact thumbs, random images posted.

**Comment by lfcnassif** [comment]
I still didn't understand how the images are related with missing status changes, could you describe with more detail as we don't have your database? 

If there is some issue here beyond status messages not supported, maybe it won't be possible to fix it without a sample database to reproduce the issue.

**Comment by gisms** [comment]
I will send you an email.

**Comment by gisms** [comment]
Final need identified:

Today the items explained below are included in Cellebrite as whatsapp chats and are not being decoded by the IPED internal parser:

a) chats created from images posted by contacts in the "Status" functionality of whatsapp (similar to instagram stories). The previewed images can be seen in a larger way, the ones that were not previewed exist only in thumbnails. In the ZWACHATSESSIONS table there are 2 different entries for the same contact, one for status (telephone@status) and another for normal conversation (telephone@s.whastapp.net) -- in cases where both exist for a particular contact.

b) Change in a contact's profile, such as change in profile picture, appear on cellebrite as a chat, without any message and the conversation has only 01 participant.

It doesn't seem like the best way to treat these two items as chats, but that's how Cellebrit does today.

**Comment by lfcnassif** [comment]
Thanks @gisms for your detailed investigation. I also don't think status and profile changes should be shown as chats (like done by Cellebrite software). But we could decode that info and show in another way...

**Comment by lfcnassif** [comment]
Filtering status changes is causing a lot of confusion among users, because seems IPED is extracting less chats than other tools, but actually they are not chats really, but contacts/owner status changes. I think we could stop filtering them, at least a proper support for status changes is implemented.

**Comment by wladimirleite** [comment]
I can try to implement proper support, as WhatsApp chats are usually a very important source of relevant information (at least nowadays).

**Comment by wladimirleite** [comment]
@lfcnassif, I checked today some databases in more detail and I have a few ideas how to implement this. 
Just "not filtering" doesn't seem simple, as all status updates (from all contacts) are stored in the database as a single chat.
One approach would be placing status updates of a given user in their chat (chronologically), with a title like "Status Update" and a different box/background color. So everything related to a contact would be in a single place.

@lfcnassif, as I mentioned before, WhatsApp related issues should have high priority, so this is something that I like to work on. 
Let me know if I can proceed and if have other ideas about how to show the status updates. I can try to implement what I have in mind and share some screenshots here, which is usually easier than trying to describe.

**Comment by lfcnassif** [comment]
Hi @wladimirleite!

> Just "not filtering" doesn't seem simple, as all status updates (from all contacts) are stored in the database as a single chat.

Hum... this is not what I expected, given PA behavior of, AFAIK, separating them per user...

> One approach would be placing status updates of a given user in their chat (chronologically), with a title like "Status Update" and a different box/background color. So everything related to a contact would be in a single place.

Seems very nice! I thought about creating a new item/chat type, per contact, and category, so the sum of items of the new category and the WhatsApp chat category would be similar to what is shown by PA, maybe avoiding user confusion when comparing results of both tools. But now I'm not sure if PA really shows status changes per contact as chats, given the database schema you described...

**Comment by wladimirleite** [comment]
> Hi [@wladimirleite](https://github.com/wladimirleite)!
> 
> > Just "not filtering" doesn't seem simple, as all status updates (from all contacts) are stored in the database as a single chat.
> 
> Hum... this is not what I expected, given PA behavior of, AFAIK, separating them per user...
> 
> > One approach would be placing status updates of a given user in their chat (chronologically), with a title like "Status Update" and a different box/background color. So everything related to a contact would be in a single place.
> 
> Seems very nice! I thought about creating a new item/chat type, per contact, and category, so the sum of items of the new category and the WhatsApp chat category would be similar to what is shown by PA, maybe avoiding user confusion when comparing results of both tools. But now I'm not sure if PA really shows status changes per contact as chats, given the database schema you described...

Separate chats is an option, but status chats will usually be overlooked (which may be a good or bad thing, depending on the point of view).
I personally find more useful to find messages and status updates in a single place, in chronological order.
I will take a closer look how it is presented in other tools.

**Comment by wladimirleite** [comment]
I just checked, and status updates are shown as separated chats in Cellebrite tools, with "Chat Type" = "Broadcast".
We can have the same behavior (maybe just with a more clear identification, like "Status Updates" in the generated HTML (maybe also a different top bar color), so one can easily identify it is not a regular chat.

I will try to implement this way and share some screenshots here.

**Comment by wladimirleite** [comment]
> (...) as all status updates (from all contacts) are stored in the database as a single chat. 

That is true for **Android** databases. 
For **iOS**, each contact with status updates has another chat, with contact identification like `xxx@status` (instead of `xxx@s.whatsapp.net`).

**Comment by wladimirleite** [comment]
Processing a few UFDRs extracted from Android devices with multiple backup databases, sometimes I got `ConcurrentModificationException`'s. 
This is kind of tricky, but I managed to reproduce processing only the WA databases several times. 
I made a simple modification that avoids the issue.
